---
title: "BowtieXEmaseTest"

author: "Sam Ardery"

date: "`r Sys.Date()`"

output: 
  html_document: 
    toc: true 
    toc_float: true 
    theme: spacelab
    code_folding: hide
---

Within the current pipeline used for this research, there are two locations where the paired-end analysis could be performed. One location is within the bowtie step using the canonical parameters for a paired-end bowtie analysis, the other being in the emase step, where two single end files are intersected and the algorithm isolates the most likely alignments. In this document, I will analyze the difference in expression results to determine if one method has more accurate results over the other.

# Packages

The packages used in this document are `here` for data file locating, `tidyverse` for data manipulation and visualization, `Hmisc` for correlation testing, `biomaRt` for genomic annotation, `ggpubr` for publication ready figure themes, and `formattable` for table design. The ensembl used is v91 from December 2017, as that was the version used when this data was initially analyzed.

```{r Load-packages, message=FALSE}
library(here)
library(tidyverse)
library(Hmisc)
library(biomaRt)
library(ggpubr)
library(assertthat)
library(eulerr)

#Creating an ensemble mart is essential to obtain the biotypes of the various filtration of genes.
ensembl <- useMart(biomart="ensembl",host="http://dec2017.archive.ensembl.org","mmusculus_gene_ensembl") #states a data set for use, in this case, the mouse data set
```

# Data

This code takes the outputs of the gbrs pipeline and formats it into a count matrix that can be used for analysis. It was provided by Dr. Selcan Aydin.

```{r Read-in-TruePE-data, message = FALSE, results = "hide"}
#Make total gene expression matrix from GBRS output
gbrsdir <- '/projects/munger-lab/ArderyProject/PE_Data'
gbrscountfiles <- list.files(gbrsdir, pattern=".diploid.genes.expected_read_counts", full=TRUE)
gbrstpmfiles <- list.files(gbrsdir, pattern=".diploid.genes.tpm",full=TRUE)
file1 <- read_tsv(gbrscountfiles[1])

read_gbrs <- function(filename) {
  samplename <- strsplit(basename(filename), "\\.")[[1]][1]
  str_replace(samplename, pattern = "_ES_", replacement = "_")
  substring(samplename, 6) <- "."
  dat <- read_tsv(filename) %>% mutate(sample=samplename)
  dat$rep <- ifelse((grepl("_repB", filename) | grepl("_B", samplename)), "repB", "repA")
  dat <- unite(dat, id, sample, rep, remove=FALSE)
  assert_that(are_equal(dat$locus, file1$locus))
  dplyr::select(dat, locus, id, sample, rep, total)
}

counts <- bind_rows(lapply(gbrscountfiles, read_gbrs)) %>%
  rename(count=total) 
tpm <- bind_rows(lapply(gbrstpmfiles, read_gbrs)) %>% 
  rename(tpm=total) 
```

```{r Filter-TPE-for-significance, message = FALSE}
# Take genes with tpm > 0.5 and >50% nonzero counts
final <- inner_join(counts, tpm) %>%
  group_by(locus) %>%
  filter(median(tpm) > 0.5, mean(count == 0) < 0.5) %>%
  dplyr::select(-tpm)
print(paste("Gene Quantity: ", length(unique(final$locus))))    # ~15k genes
```

This chunk takes the count data from above and reformats it to allow for easy analysis and legibility.

```{r Reformat-TPE-data}
#isolate the 3rd and fourth columns which contain the sample ids, the gene ids, and the counts
TruePE <- final[,-c(3:4)]
#remove the extraneous bits from the id
TruePE$id <- substr(TruePE$id, 1,8)
TruePE$id <- paste(TruePE$id, final$rep, sep = "_")
#round the count variable to four decimal places
TruePE$count <- format(round(TruePE$count, 4))
#reformat the table to make every sample a single column.
TruePE <- pivot_wider(TruePE, names_from = id, values_from = count)
TruePE <- dplyr::rename(TruePE, GeneID = locus)
```

This is the pseudo paired-end data, where the PE sequencing files were aligned as two single end files which were then intersected using the EMASE algorithm.

```{r Load-PseudoPE-data}
# load the pe - paired end data
pe_data <- load(here("SideComparisons", "DO185_mESC_paired_emase_m4_noProbs_correctedIDs_02012021.RData"))
PseudoPE <- as.data.frame(expr) %>% rownames_to_column("GeneID") %>% as_tibble() #PE = paired-end
rm(DO185_mESC_emase_m4.cmd, raw.expr, covar, covarTidy, matches, exprNotes, pe_data, mixupNotes, expr)
```

Creating matched data sets allows for ease of analysis later on in the document. The use of the semi join function means that each data set only contains ids that are found in both data sets but they are still associated only with the values of each respective alignment method. I will also create another joined data frame, this one containing the same gene lists as the semi_joined data frames but also containing the counts from both alignment methods in one data frame.

```{r Create-matched-data-tables}
matchedTPE <- semi_join(TruePE, PseudoPE, by = "GeneID")
matchedPPE <- semi_join(PseudoPE, TruePE, by = "GeneID")

Comp <- inner_join(TruePE, PseudoPE, by = "GeneID")
```

Loading the single-end data. This will be used for part of the analysis later on.

```{r Load-single-end-data}
# load the R1 - single read data 
r1_data <- load(here("SideComparisons", "DO185_mESC_emase_m4_noProbs_correctedIDs_02012021.RData"))
esc_expr_r1 <- expr # cleaned, normalized, single read, R1
esc_raw_expr_r1 <- raw.expr # raw data, only filtered not normalized
rm(matches, exprNotes, mixupNotes, DO185_mESC_emase_m4.cmd, expr, raw.expr, covar, covarTidy) #removing necessary data

SE <- as.data.frame(esc_expr_r1) %>% rownames_to_column("GeneID") %>% as_tibble()
```

# Analysis {.tabset .tabset-pills}

## Correlation
The first step is to compare the gene counts provided by each alignment methods between the same samples. This is done in the same way as the Expression Comparison document. The pearson and the spearman correlation is determined for each sample comparison and the entire collection of statistics are then able to be analyzed.

```{r Gene-comparison}
pearsonGenes <- c() #This vector will get the pearson correlation coefficient for each iteration of the loop
spearmanGenes <- c()
GeneID <- c() #This vector will hold the gene IDs for each loop

#This loop will perform the above comparison on all of the genes.
for(i in 1:14179) {
  GeneTPE <- matchedTPE[i, ]
  GeneTPE <- pivot_longer(data = GeneTPE,
                         cols = -GeneID,
                         names_to = "Sample",
                         values_to = "TPE")
  GenePPE <- matchedPPE[i, ]
  GenePPE <- pivot_longer(data = GenePPE,
                         cols = -GeneID,
                         names_to = "Sample",
                         values_to = "PPE")
  name <- GeneTPE$GeneID[1]
  Join_holder <- left_join(GeneTPE, GenePPE, by = c("Sample", "GeneID"))
  #Join_holder$TPE <- replace_na(Join_holder$TPE, -1)
  #Join_holder$PPE <- replace_na(Join_holder$PPE, -1)
  #Join_holder <- dplyr::select(Join_holder, 2:4)

  GeneID <- append(GeneID, name) #adding the gene name to a pre-made vector
  peaGenes <- rcorr(Join_holder$TPE, Join_holder$PPE, type = "pearson") #determining the pearson correlation coefficient
  pearsonGenes <- append(pearsonGenes, peaGenes$r[3]) #adding the correlation coefficient to a pre-made vector
  spearGenes <- rcorr(pull(Join_holder, 3), pull(Join_holder, 4), type = "spearman")
  spearmanGenes <- append(spearmanGenes, spearGenes$r[3])
}
PearsonGenes <- data.frame(GeneID, pearsonGenes) #combining the two previous vectors into one data frame
SpearmanGenes <- data.frame(GeneID, spearmanGenes)

summary(PearsonGenes)
summary(SpearmanGenes)
```

```{r Plot-gene-correlations}
#plotting the pearson correlations distribution
ggplot(PearsonGenes, aes(x = pearsonGenes), pty = "s") +
  geom_histogram(binwidth = 0.01) +
  labs(title = "Gene Correlation Distribution Pseudo vs True", x = "Pearson Correlation Coefficient", y = "Frequency") +
  theme_pubr()

#plotting the spearman correlation distribution
ggplot(SpearmanGenes, aes(x = spearmanGenes), pty = "s") +
  geom_histogram(binwidth = 0.01) +
  labs(title = "Gene Correlation Distribution Pseudo vs True", x = "Spearman Correlation Coefficient", y = "Frequency") +
  theme_pubr()
```

## Biotype

In order to get a better idea of what the reads are actually aligning to in each method, the biotype of each gene that has reads is determined. This is done with the Comp table, allowing for the joining of the table with the representative biotypes.

```{r Biotype-analysis}
#obtaining the biotype for each gene id in the data set
CompBiotypes <- getBM(attributes = c('ensembl_gene_id', 'gene_biotype'), filters = 'ensembl_gene_id', values = Comp$GeneID, mart = ensembl)
#renaming the resulting table names to aid in table joining
CompBiotypes <- dplyr::rename(CompBiotypes, GeneID = ensembl_gene_id)

#joining the expression data with the biotype data and grouping by biotype
ExpXBiotype <- full_join(Comp, CompBiotypes, by = "GeneID")
ExpXBiotype <- group_by(ExpXBiotype, gene_biotype)

#Determining the frequency of each biotype in the data
ExpXBiotype %>%
  count(gene_biotype) -> BiotypeCounts

#plotting the frequency of each biotype
ggplot(data = BiotypeCounts, aes(x = n, y = gene_biotype)) +
  geom_col() +
  labs(title = "Combined Biotypes",x = "Frequency", y = "GeneBiotype")

#plotting the frequency of each biotype excluding the protein coding genes
BiotypeCounts %>%
  filter(gene_biotype != "protein_coding") %>%
  ggplot(aes(x = n, y = gene_biotype)) +
    geom_col() +
    labs(title = "Combined Biotypes w/o protein coding", x = "Frequency", y = "Gene Biotype") 
```

## Unique

In order to understand the differences between each alignment method, we have to isolate the unique data provided in each respective data table. We can then get the biotypes and further analyze if one method is improved over the other in these biotype categories.

```{r Unique-data-analysis}
#creating a table that contains gene ids only identified via true PE alignment
UTruePE <- anti_join(TruePE, PseudoPE, by = "GeneID")
#determining the biotypes for these gene ids
UBBiotype <- getBM(attributes = c('ensembl_gene_id', 'gene_biotype'), filters = 'ensembl_gene_id', values = UTruePE$GeneID, mart = ensembl)
UBBiotype <- dplyr::rename(UBBiotype, GeneID = ensembl_gene_id)
#joining the expression and biotype data for the unique True PE data 
UBExpBio <- full_join(UBBiotype, UTruePE, by = "GeneID")
#determining and plotting the biotype frequency 
UBCounts <- count(UBBiotype, gene_biotype)
ggplot(data = UBCounts, aes(x = n, y = gene_biotype)) +
  geom_col() +
  labs(title = "Unique TruePE Biotypes",x = "Frequency", y = "Biotypes") +
  theme_pubr()

#the same will be done for the pseudo PE data 

#creating a table that contains gene ids only identified via true PE alignment
UPseudoPE <- anti_join(PseudoPE, TruePE, by = "GeneID")
#determining the biotypes for these gene ids
UEBiotype <- getBM(attributes = c('ensembl_gene_id', 'gene_biotype'), filters = 'ensembl_gene_id', values = UPseudoPE$GeneID, mart = ensembl)
UEBiotype <- dplyr::rename(UEBiotype, GeneID = ensembl_gene_id)
#joining the expression and biotype data for the unique Pseudo PE data 
UEExpBio <- full_join(UEBiotype, UPseudoPE, by = "GeneID")
#determining and plotting the biotype frequency
UECounts <- count(UEBiotype, gene_biotype)
ggplot(data = UECounts, aes(x = n, y = gene_biotype)) +
  geom_col() +
  labs(title = "Unique PseudoPE Biotypes",x = "Frequency", y = "Biotypes") +
  theme_pubr()
```

## PseudoPE x SE Comparison

Here is a brief comparison of the change in expression levels between the PseudoPE alignment and the SE alignment. This was done for one gene but is representative of most of the genes in the data set. This shows the increase in expression counts in single-end alignment.

```{r PseudoPE-by-SE-stripcharts}
#filtering for the gene in the PseudoPE data set and pivoting the table
teststripPE <- filter(PseudoPE, GeneID == "ENSMUSG00000052429")
teststripPE <- pivot_longer(data = teststripPE, cols = c(2:185), names_to = "Sample", values_to = "PseudoPE") 
#filtering for the gene in the SE data set and pivoting the table
teststripSE <- filter(SE, GeneID == "ENSMUSG00000052429")
teststripSE <- pivot_longer(data = teststripSE, cols = c(2:185), names_to = "Sample", values_to = "SE") 
#joining the two tables   
teststrip <- full_join(teststripPE, teststripSE, by = c("GeneID" = "GeneID", "Sample" = "Sample"))
#creating a list of the lists of data 
x <- list('Pseudo' = teststrip$PseudoPE, 'SE' = teststrip$SE)

#plotting the data in a strip chart with lines connecting the same sample between the data sets
stripchart(x, vertical = TRUE, group.names = c("PseudoPE", "SE"), pch = 19, col = "green", main = "Expression of ENSMUSG00000052429") 
for(i in 1:nrow(teststrip)) {segments(1, x$Pseudo[i], 2, x$SE[i])}
```

## Rps7 and Pseudogene Gm9493

In the initial analysis of this data, Rps7 was identified as a gene with high ability to mis-align between single-end and paired-end. Here, we continue the analysis of this pair of genes Rps7 = ENSMUSG00000061477 Gm9493 = ENSMUSG00000062353

```{r Rps7-comparison}
#filtering True paired end for the two genes and pivoting the data 
TruePE %>%
  filter(GeneID == "ENSMUSG00000061477" | GeneID == "ENSMUSG00000062353") %>%
  pivot_longer(cols = c(2:169), names_to = "Sample", values_to = "TruePE") -> TrueRPS

#filtering pseudo paired end for the two genes and pivoting the data 
PseudoPE %>%
  filter(GeneID == "ENSMUSG00000061477" | GeneID == "ENSMUSG00000062353") %>%
  pivot_longer(cols = c(2:170), names_to = "Sample", values_to = "PseudoPE") -> PseudoRPS

#joining the true and pseudo Rps7/Gm9493 data
CompRPS <- full_join(TrueRPS, PseudoRPS, by = c("GeneID" = "GeneID", "Sample" = "Sample"))
#replacing NA values with -1 in the True and Pseudo columns
CompRPS$TruePE <- as.numeric(CompRPS$TruePE)
CompRPS$TruePE <- replace_na(CompRPS$TruePE, -1)
CompRPS$PseudoPE <- replace_na(CompRPS$PseudoPE, -1)
#converting the TruePE values to numeric
CompRPS$TruePE <- as.numeric(CompRPS$TruePE)

#plotting the comparison
ggplot(CompRPS, aes(x = TruePE, y = PseudoPE, color = GeneID)) +
  geom_point() +
  scale_x_continuous(limits = c(-5,75000)) +
  scale_y_continuous(limits = c(-5,75000)) +
  labs(title = "Rps7 and Pseudogene Comparison", x = "TruePE", y = "PseudoPE") 
```

## Venn Diagrams

To clearly present the differences between each sample in terms of the biotypes, a Venn diagram can be created to show the differences in quantity of a specific biotype. This is first done with the pseudogenes, allowing us to see which alignment method identifies more pseudogenes(quantity of pseudogenes is presently used as a metric for accuracy in this document, with higher quantities of pseudogenes relating to lower accuracy).

```{r Pseudogene-venn-diagram}
#determine the biotypes for the full True PE data set
TruePEBiotypes <- getBM(attributes = c('ensembl_gene_id', 'gene_biotype'), filters = 'ensembl_gene_id', values = TruePE$GeneID, mart = ensembl)
#filter for pseudogenes
TruePEPseudogenes <- filter(TruePEBiotypes, gene_biotype == "processed_pseudogene")

#determine the biotypes for the full PseudoPE data set
PseudoPEBiotypes <- getBM(attributes = c('ensembl_gene_id', 'gene_biotype'), filters = 'ensembl_gene_id', values = PseudoPE$GeneID, mart = ensembl)
#filter for pseudogenes
PseudoPEPseudogenes <- filter(PseudoPEBiotypes, gene_biotype == "processed_pseudogene")

#determine the biotype for the single-end data set
SEBiotypes <- getBM(attributes = c('ensembl_gene_id', 'gene_biotype'), filters = 'ensembl_gene_id', values = SE$GeneID, mart = ensembl)
#filter for pseudogenes
SEPseudogenes <- filter(SEBiotypes, gene_biotype == "processed_pseudogene")

#use the pseudogene lists above to create a list of lists of the pseudogene gene ids
Venn <- list(True = TruePEPseudogenes$ensembl_gene_id, Pseudo = PseudoPEPseudogenes$ensembl_gene_id, SE = SEPseudogenes$ensembl_gene_id)

#plot the euler diagram (circles are sized to show relative quantities) with the quantity shown on the diagram
plot(euler(Venn), quantities = TRUE)
```

This can also be done for protein coding genes, which can also be used as a metric for accuracy, with higher protein coding quantities relating to higher accuracy.

```{r Protein-coding-venn-diagram}
#determine the biotypes for the full True PE data set
TruePEBiotypes <- getBM(attributes = c('ensembl_gene_id', 'gene_biotype'), filters = 'ensembl_gene_id', values = TruePE$GeneID, mart = ensembl)
#filter for pseudogenes
TruePEProteins <- filter(TruePEBiotypes, gene_biotype == "protein_coding")

#determine the biotypes for the full PseudoPE data set
PseudoPEBiotypes <- getBM(attributes = c('ensembl_gene_id', 'gene_biotype'), filters = 'ensembl_gene_id', values = PseudoPE$GeneID, mart = ensembl)
#filter for pseudogenes
PseudoPEProteins <- filter(PseudoPEBiotypes, gene_biotype == "protein_coding")

#determine the biotype for the single-end data set
SEBiotypes <- getBM(attributes = c('ensembl_gene_id', 'gene_biotype'), filters = 'ensembl_gene_id', values = SE$GeneID, mart = ensembl)
#filter for pseudogenes
SEProteins <- filter(SEBiotypes, gene_biotype == "protein_coding")

#use the protein coding lists above to create a list of lists of the protein coding gene ids
Venn <- list(True = TruePEProteins$ensembl_gene_id, Pseudo = PseudoPEProteins$ensembl_gene_id, SE = SEProteins$ensembl_gene_id)

#plot the euler diagram (circles are sized to show relative quantities) with the quantity shown on the diagram
plot(euler(Venn), quantities = TRUE)
#plot the data as a venn diagram to more clearly show the quantities
plot(venn(Venn), quantities = TRUE)
```
